File: src/dependencies/crazyflie_hardware/src/crtp/crtp/logic/toc_logic.py
from .logic import Logic
from crtp.crtp_link import CrtpLink
from .toc.toc import Toc
from .toc.toccache import TocCache
from crtp.packers.toc_packer import TocPacker

import struct


class TocLogic(Logic):

    def __init__(
        self,
        packer: TocPacker,  # crtp_packer_factory: Callable[[int], CrtpPacker],
        crtp_link: CrtpLink,
        ElementType,
        path: str,
    ):
        super().__init__(crtp_link)
        self.packer: TocPacker = packer
        self.ElementType = ElementType

        self.toc = Toc()
        self.toc_cache = TocCache(rw_cache=path)

        self.nbr_of_items = None
        self.crc = None

    def _to_toc_item(self, data):
        ident = struct.unpack("<H", data[1:3])[0]
        data = bytearray(data[3:])
        element = self.ElementType(ident, data)
        self.toc.add_element(element)

    def initialize_toc(self):
        nbr_of_items, crc = self.send_get_toc_info()
        cache_data = self.toc_cache.fetch(crc)
        if cache_data:
            self.toc.toc = cache_data
        else:
            self.send_download_toc_items()

    def send_get_toc_info(self):
        packet, expects_response, matching_bytes = self.packer.get_toc_info()
        resp_data = self.link.send_packet(packet, expects_response, matching_bytes).data
        [self.nbr_of_items, self.crc] = struct.unpack("<HI", resp_data[1:7])
        return self.nbr_of_items, self.crc

    def send_download_toc_items(self):
        if self.nbr_of_items == None or self.crc == None:
            self.send_get_toc_info()
        packets = []
        for i in range(self.nbr_of_items):
            packets.append(self.packer.get_toc_item(i))

        responses = self.link.send_batch_request(packets)
        for result in responses:
            self._to_toc_item(result.packet.data[: result.packet.data_length])

        self.toc_cache.insert(self.crc, self.toc.toc)

File: src/dependencies/crazyflie_hardware/src/crtp_cpp/src/crtp_cpp/logic/toc_logic.cpp
#include "crtp_cpp/logic/toc_logic.hpp"
#include <stdexcept>
#include <cstring>
#include <fstream>

#include "crtp_cpp/logic/parameters_logic.hpp"
#include "crtp_cpp/logic/logging_logic.hpp"

template <class T>
TocLogic<T>::TocLogic(CrtpLink *crtp_link, const std::string &path, uint8_t port)
    : Logic(crtp_link), toc_cache_path(path), packer(TocPacker(port)) {}

template <class T>
bool TocLogic<T>::load_from_file(uint32_t crc)
{
    std::string fileName = std::to_string(crc) + ".csv";
    std::ifstream infile(fileName);
    if (!infile.good())
        return false;

    toc_entries.clear();
    std::string line;
    while (std::getline(infile, line))
    {
        toc_entries.push_back(T(line));
    }
    return true;
}

template <class T>
void TocLogic<T>::write_to_file()
{
    if (!nbr_of_items.has_value() || !crc.has_value())
    {
        send_download_toc_items();
    }

    std::string fileName = std::to_string(crc.value()) + ".csv";
    std::string fileNameTemp = fileName + ".tmp";
    std::ofstream output(fileNameTemp);
    for (const auto &entry : toc_entries)
    {
        output << entry.toString() << std::endl;
    }
    // change the filename
    rename(fileNameTemp.c_str(), fileName.c_str());
}

template <class T>
void TocLogic<T>::initialize_toc()
{
    auto [nbr_of_items, crc] = send_get_toc_info();
    bool cached = load_from_file(crc);
    if (!cached)
    {
        send_download_toc_items();
        write_to_file();
    }
}

template <class T>
std::pair<uint16_t, uint32_t> TocLogic<T>::send_get_toc_info()
{
    CrtpRequest request = packer.get_toc_info();
    auto response = link->send_packet(request);

    uint16_t nbr_of_items;
    uint32_t crc;
    if (!response || response.value().data_length < 7)
    {
        nbr_of_items = 0;
        crc = 0;
        // throw std::runtime_error("Invalid TOC info response");
    }
    else
    {
        std::memcpy(&nbr_of_items, response.value().data + 1, sizeof(uint16_t));
        std::memcpy(&crc, response.value().data + 3, sizeof(uint32_t));
    }

    this->nbr_of_items = nbr_of_items;
    this->crc = crc;
    return {nbr_of_items, crc};
}

template <class T>
void TocLogic<T>::send_download_toc_items()
{
    if (!nbr_of_items.has_value() || !crc.has_value())
    {
        send_get_toc_info();
    }

    std::vector<CrtpRequest> requests;
    for (uint16_t i = 0; i < nbr_of_items.value(); ++i)
    {
        requests.push_back(packer.get_toc_item(i));
    }

    toc_entries.clear();
    auto responses = link->send_batch_request(requests);
    for (const auto &packet : responses)
    {
        std::vector<uint8_t> data(packet.data, packet.data + packet.data_length);
        toc_entries.push_back(T(data));
    }
}

template class TocLogic<ParamTocEntry>;
template class TocLogic<LogTocEntry>;